File: rust/candid_parser/tests/assets/ok/fieldnat.js
import { IDL } from '@dfinity/candid';

export const tuple = IDL.Tuple(IDL.Text, IDL.Text);
export const non_tuple = IDL.Record({ _1_ : IDL.Text, _2_ : IDL.Text });

export const idlService = IDL.Service({
  'bab' : IDL.Func([IDL.Int, IDL.Nat], [], []),
  'bar' : IDL.Func(
      [IDL.Record({ '2' : IDL.Int })],
      [IDL.Variant({ 'e20' : IDL.Null, 'e30' : IDL.Null })],
      [],
    ),
  'bas' : IDL.Func(
      [IDL.Tuple(IDL.Int, IDL.Int)],
      [IDL.Tuple(IDL.Text, IDL.Nat)],
      [],
    ),
  'baz' : IDL.Func(
      [IDL.Record({ _2_ : IDL.Int, '2' : IDL.Nat })],
      [IDL.Record({})],
      [],
    ),
  'bba' : IDL.Func([tuple], [non_tuple], []),
  'bib' : IDL.Func([IDL.Tuple(IDL.Int)], [IDL.Variant({ _0_ : IDL.Int })], []),
  'foo' : IDL.Func(
      [IDL.Record({ _2_ : IDL.Int })],
      [IDL.Record({ _2_ : IDL.Int, '_2' : IDL.Int })],
      [],
    ),
});

export const idlInitArgs = [];

/**
 * @deprecated Import IDL types directly from this module instead of using this factory function.
 */
export const idlFactory = ({ IDL }) => {
  const tuple = IDL.Tuple(IDL.Text, IDL.Text);
  const non_tuple = IDL.Record({ _1_ : IDL.Text, _2_ : IDL.Text });
  return IDL.Service({
    'bab' : IDL.Func([IDL.Int, IDL.Nat], [], []),
    'bar' : IDL.Func(
        [IDL.Record({ '2' : IDL.Int })],
        [IDL.Variant({ 'e20' : IDL.Null, 'e30' : IDL.Null })],
        [],
      ),
    'bas' : IDL.Func(
        [IDL.Tuple(IDL.Int, IDL.Int)],
        [IDL.Tuple(IDL.Text, IDL.Nat)],
        [],
      ),
    'baz' : IDL.Func(
        [IDL.Record({ _2_ : IDL.Int, '2' : IDL.Nat })],
        [IDL.Record({})],
        [],
      ),
    'bba' : IDL.Func([tuple], [non_tuple], []),
    'bib' : IDL.Func(
        [IDL.Tuple(IDL.Int)],
        [IDL.Variant({ _0_ : IDL.Int })],
        [],
      ),
    'foo' : IDL.Func(
        [IDL.Record({ _2_ : IDL.Int })],
        [IDL.Record({ _2_ : IDL.Int, '_2' : IDL.Int })],
        [],
      ),
  });
};

/**
 * @deprecated Import IDL types directly from this module instead of using this factory function.
 */
export const init = ({ IDL }) => { return []; };

File: rust/candid_parser/tests/assets/ok/fieldnat.rs
// This is an experimental feature to generate Rust binding from Candid.
// You may want to manually adjust some of the types.
#![allow(dead_code, unused_imports)]
use candid::{self, CandidType, Deserialize, Principal};
use ic_cdk::call::{Call, CallResult as Result};

#[derive(CandidType, Deserialize)]
pub struct BarArg { #[serde(rename="2")] pub _50_: candid::Int }
#[derive(CandidType, Deserialize)]
pub enum BarRet { #[serde(rename="e20")] E20, #[serde(rename="e30")] E30 }
#[derive(CandidType, Deserialize)]
pub struct BazArg {
  pub _2_: candid::Int,
  #[serde(rename="2")]
  pub _50_: candid::Nat,
}
#[derive(CandidType, Deserialize)]
pub struct BazRet {}
#[derive(CandidType, Deserialize)]
pub struct Tuple (pub String, pub String);
#[derive(CandidType, Deserialize)]
pub struct NonTuple { pub _1_: String, pub _2_: String }
#[derive(CandidType, Deserialize)]
pub enum BibRet { _0_(candid::Int) }
#[derive(CandidType, Deserialize)]
pub struct FooArg { pub _2_: candid::Int }
#[derive(CandidType, Deserialize)]
pub struct FooRet { pub _2_: candid::Int, pub _2: candid::Int }

pub struct Service(pub Principal);
impl Service {
  pub async fn bab(&self, two: &candid::Int, arg1: &candid::Nat) -> Result<()> {
    Ok(Call::bounded_wait(self.0, "bab").with_args(&(two,arg1,)).await?.candid()?)
  }
  pub async fn bar(&self, arg0: &BarArg) -> Result<(BarRet,)> {
    Ok(Call::bounded_wait(self.0, "bar").with_args(&(arg0,)).await?.candid()?)
  }
  pub async fn bas(&self, arg0: &(candid::Int, candid::Int)) -> Result<((String, candid::Nat),)> {
    Ok(Call::bounded_wait(self.0, "bas").with_args(&(arg0,)).await?.candid()?)
  }
  pub async fn baz(&self, arg0: &BazArg) -> Result<(BazRet,)> {
    Ok(Call::bounded_wait(self.0, "baz").with_args(&(arg0,)).await?.candid()?)
  }
  pub async fn bba(&self, arg0: &Tuple) -> Result<(NonTuple,)> {
    Ok(Call::bounded_wait(self.0, "bba").with_args(&(arg0,)).await?.candid()?)
  }
  pub async fn bib(&self, arg0: &(candid::Int)) -> Result<(BibRet,)> {
    Ok(Call::bounded_wait(self.0, "bib").with_args(&(arg0,)).await?.candid()?)
  }
  pub async fn foo(&self, arg0: &FooArg) -> Result<(FooRet,)> {
    Ok(Call::bounded_wait(self.0, "foo").with_args(&(arg0,)).await?.candid()?)
  }
}
/// Canister ID: `aaaaa-aa`
pub const CANISTER_ID : Principal = Principal::from_slice(&[]);
pub const service : Service = Service(CANISTER_ID);

